src/rendergraph/opengl/backend/basegeometry.h

```diff
@@ -0,0 +1,56 @@
+#pragma once
+
+#include <QString>
+#include <vector>
+
+#include "rendergraph/types.h"
+
+namespace rendergraph {
+class BaseAttributeSet; // fwd decl to avoid circular dependency
+class BaseGeometry;
+}
+
+// TODO this assumes all vertices consist of floats
+class rendergraph::BaseGeometry {
+  protected:
+    BaseGeometry(const BaseAttributeSet& attributeSet, int vertexCount);
+
+  public:
+    struct Attribute {
+        const int m_offset;
+        const int m_tupleSize;
+        const PrimitiveType m_primitiveType;
+        const QString m_name;
+    };
+
+    float* vertexData() {
+        return m_vertexData.data();
+    }
+    const float* vertexData() const {
+        return m_vertexData.data();
+    }
+    const Attribute* attributes() const {
+        return m_pAttributes;
+    }
+    int attributeCount() const {
+        return m_attributeCount;
+    }
+    int vertexCount() const {
+        return m_vertexCount;
+    }
+    int sizeOfVertex() const { // in bytes
+        return m_sizeOfVertex;
+    }
+    void allocate(int vertexCount) {
+        m_vertexCount = vertexCount;
+        m_vertexData.resize(m_vertexCount * sizeOfVertex() / sizeof(float));
+    }
+
+  protected:
+    const Attribute* m_pAttributes;
+    const int m_attributeCount;
+    const int m_sizeOfVertex;
+    int m_drawingMode;
```

Comment:
why `int` instead of the `enum`?

Reply:
For parity with QSGNode.

Reply:
can you elaborate?

Reply:
sorry, QSGGeometry.

QSGGeometry stores the drawing mode as an int, and I has a setDrawingMode in the common layer which therefore had to use an int. But I realize this is not the case anymore so I can indeed use an enum in the opengl layer. Will do!